service/storage_service.cc

```diff
@@ -940,39 +940,61 @@ future<> storage_service::handle_state_normal(inet_address endpoint) {
     // the collision is resolved by assigning the token to the endpoint which started later.
     std::unordered_set<token> owned_tokens;
 
+    // token_to_endpoint_map is used to track the current token owners.
+    // when any tokne is replaced by a new owner, we track the existing owner in `candidates_for_removal`
+    // and eventually, if any candidate for removal end up owning no tokens, it is removed
+    // from token_metadata.
+    std::unordered_map<token, inet_address> token_to_endpoint_map = get_token_metadata().get_token_to_endpoint();
+    std::unordered_set<inet_address> candidates_for_removal;
+
     for (auto t : tokens) {
         // we don't want to update if this node is responsible for the token and it has a later startup time than endpoint.
-        auto current_owner = tmptr->get_endpoint(t);
-        if (!current_owner) {
+        auto current = token_to_endpoint_map.find(t);
+        if (current == token_to_endpoint_map.end()) {
             slogger.debug("handle_state_normal: New node {} at token {}", endpoint, t);
             owned_tokens.insert(t);
-        } else if (endpoint == *current_owner) {
-            slogger.debug("handle_state_normal: endpoint={} == current_owner={} token {}", endpoint, *current_owner, t);
+            continue;
+        }
+        auto current_owner = current->second;
+        if (endpoint == current_owner) {
+            slogger.debug("handle_state_normal: endpoint={} == current_owner={} token {}", endpoint, current_owner, t);
             // set state back to normal, since the node may have tried to leave, but failed and is now back up
             owned_tokens.insert(t);
-        } else if (_gossiper.compare_endpoint_startup(endpoint, *current_owner) > 0) {
-            slogger.debug("handle_state_normal: endpoint={} > current_owner={}, token {}", endpoint, *current_owner, t);
+            token_to_endpoint_map.erase(current);
```

Comment:
Why this `erase`? `current_owner` is still the owner in this case.

IIUC it shouldn't matter for correctness, we will not insert `endpoint` into `candidates_for_removal`, so even if we `erase`d all of `endpoint`'s tokens from `token_to_endpoint_map`, it won't cause `endpoint` to end up in `endpoints_to_remove`.

But it does make things a bit confusing, it goes against the definition of `token_to_endpoint_map` which you gave in the comment, i.e. it tracks the current token owners.

Reply:
Yes, it's doesn't matter for correctness.
It's just an optimization for speeding up the traversal of token_to_endpoint_map down below
I can delete this if you think it's confusing.